pkg/services/accesscontrol/filter.go

```diff
@@ -3,101 +3,64 @@ package accesscontrol
 import (
 	"context"
 	"errors"
-	"fmt"
+	"strconv"
 	"strings"
 
 	"github.com/grafana/grafana/pkg/models"
-	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
 )
 
 var sqlIDAcceptList = map[string]struct{}{
 	"org_user.user_id": {},
 }
 
-type SQLDialect interface {
-	DriverName() string
-}
+const denyQuery = " 1 = 0"
 
 // Filter creates a where clause to restrict the view of a query based on a users permissions
 // Scopes for a certain action will be compared against prefix:id:sqlID where prefix is the scope prefix and sqlID
 // is the id to generate scope from e.g. user.id
-func Filter(ctx context.Context, dialect SQLDialect, sqlID, prefix, action string, user *models.SignedInUser) (string, []interface{}, error) {
+func Filter(ctx context.Context, sqlID, prefix, action string, user *models.SignedInUser) (string, []interface{}, error) {
 	if _, ok := sqlIDAcceptList[sqlID]; !ok {
-		return "", nil, errors.New("sqlID is not in the accept list")
+		return denyQuery, nil, errors.New("sqlID is not in the accept list")
 	}
-
 	if user.Permissions == nil || user.Permissions[user.OrgId] == nil {
-		return "", nil, errors.New("missing permissions")
+		return denyQuery, nil, errors.New("missing permissions")
 	}
 
-	scopes := user.Permissions[user.OrgId][action]
-	if len(scopes) == 0 {
-		return " 1 = 0", nil, nil
+	var hasWildcard bool
+	var ids []interface{}
+	for _, scope := range user.Permissions[user.OrgId][action] {
+		if strings.HasPrefix(scope, prefix) {
+			if scope[len(scope)-1] == '*' {
+				hasWildcard = true
+				break
+			}
+			if id, err := parseScopeID(scope); err == nil {
+				ids = append(ids, id)
+			}
+		}
 	}
 
-	var sql string
-	var args []interface{}
-
-	switch {
-	case strings.Contains(dialect.DriverName(), migrator.SQLite):
-		sql, args = sqliteQuery(scopes, sqlID, prefix)
-	case strings.Contains(dialect.DriverName(), migrator.MySQL):
-		sql, args = mysqlQuery(scopes, sqlID, prefix)
-	case strings.Contains(dialect.DriverName(), migrator.Postgres):
-		sql, args = postgresQuery(scopes, sqlID, prefix)
-	default:
-		return "", nil, fmt.Errorf("unknown database: %s", dialect.DriverName())
+	if hasWildcard {
+		return " 1 = 1", nil, nil
```

Comment:
nit: since deny query has a constant, this one could also have it

Reply:
:+1:

:+1: counterarg to that is that deny is used three times, this is only used once, so the need for a constant is lower

I think we could go both ways, I don't really like having to jump around to validate that the strings do what they claim they do, but there's also always the risk of typos 🤔